### PR TITLE
account: EnsureTokenIdCacheAsync

### LINKED ISSUE #2547: Missing HLE command 'EnsureIdTokenCacheAsync' in 'IManagerForApplication' (AM)
Super Meat boy gets stuck on launching page without the update 1.0.4
[yuzu_log before update.txt](https://github.com/yuzu-emu/yuzu/files/3257957/yuzu_log.before.update.txt)

still gets stuck on launching page after the update 1.0.4 but windows freeze too
[yuzu_log after update.txt](https://github.com/yuzu-emu/yuzu/files/3257964/yuzu_log.after.update.txt)

By the way,i cut out some thousands of the last lines since they were the same and i couldnt upload them(>10mb).

<meta name="titleids"  content="01004F8006A78000">

### PR DESCRIPTION
**Pull request by ogniK5377**
Closes #2547, #6946, #4239 

Death Coming now boots:
2021-09-21_14-01-d2f5dcc1-8c1c-42c8-b86d-5be4e8ec1b5f-AkjrNyWW

### CODE REVIEW

**Comment by german77** [comment]
Oh nice. I tried to fix this game the other day it just rendered black :D

**Comment by german77** [comment]
I tried with the artifact from this PR. The game still renders black. I'm starting to think my implementation was fine. It's probably AMD being AMD. I will try with NVidia latter today

File: src/core/hle/service/acc/async_context.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace Service::Account {
+IAsyncContext::IAsyncContext(Core::System& system_)
+    : ServiceFramework{system_, "IAsyncContext"}, compeletion_event{system_.Kernel()} {
+
+    Kernel::KAutoObject::Create(std::addressof(compeletion_event));
+    compeletion_event.Initialize("IAsyncContext:CompletionEvent");
+
+    // clang-format off
+        static const FunctionInfo functions[] = {
```
Comment (ameerj) [inline src/core/hle/service/acc/async_context.cpp:17]: There's an extra indent on this block

File: src/core/hle/service/acc/async_context.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    void GetResult(Kernel::HLERequestContext& ctx);
+
+protected:
+    virtual bool IsComplete() = 0;
+    virtual void Cancel() = 0;
+    virtual ResultCode GetResult() = 0;
+
+    void MarkComplete();
+
+    bool is_complete{false};
```
Comment (ameerj) [inline src/core/hle/service/acc/async_context.h:32]: Should this be made atomic?

File: src/core/hle/service/acc/async_context.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+public:
+    explicit IAsyncContext(Core::System& system_);
+
+    void GetSystemEvent(Kernel::HLERequestContext& ctx);
+    void Cancel(Kernel::HLERequestContext& ctx);
+    void HasDone(Kernel::HLERequestContext& ctx);
+    void GetResult(Kernel::HLERequestContext& ctx);
+
+protected:
+    virtual bool IsComplete() = 0;
```
Comment (lioncash) [inline src/core/hle/service/acc/async_context.h:26]:
```suggestion
    virtual bool IsComplete() const = 0;
```

File: src/core/hle/service/acc/acc.cpp [hunk 4]
```diff
@@ -504,16 +489,42 @@ class IGuestLoginRequest final : public ServiceFramework<IGuestLoginRequest> {
...
+    ResultCode GetResult() override {
+        return ResultSuccess;
+    }
+};
+
 class IManagerForApplication final : public ServiceFramework<IManagerForApplication> {
 public:
     explicit IManagerForApplication(Core::System& system_, Common::UUID user_id_)
-        : ServiceFramework{system_, "IManagerForApplication"}, user_id{user_id_} {
+        : ServiceFramework{system_, "IManagerForApplication"}, user_id{user_id_}, system(system_) {
```
Comment (Morph1984) [inline src/core/hle/service/acc/acc.cpp:521]:
```suggestion
        : ServiceFramework{system_, "IManagerForApplication"}, user_id{user_id_} {
```
system is already a protected member of ServiceFramework

File: src/core/hle/service/acc/acc.cpp [hunk 5]
```diff
@@ -504,16 +489,42 @@ class IGuestLoginRequest final : public ServiceFramework<IGuestLoginRequest> {
     }
 };
 
+class EnsureTokenIdCacheAsyncInterface final : public IAsyncContext {
+public:
+    explicit EnsureTokenIdCacheAsyncInterface(Core::System& system_) : IAsyncContext(system_) {
```
Comment (Morph1984) [inline src/core/hle/service/acc/acc.cpp:494]:
```suggestion
    explicit EnsureTokenIdCacheAsyncInterface(Core::System& system_) : IAsyncContext{system_} {
```

File: src/core/hle/service/acc/acc.cpp [hunk 6]
```diff
@@ -522,6 +533,8 @@ class IManagerForApplication final : public ServiceFramework<IManagerForApplicat
         // clang-format on
 
         RegisterHandlers(functions);
+
+        ensure_token_id = std::make_shared<EnsureTokenIdCacheAsyncInterface>(system);
```
Comment (Morph1984) [inline src/core/hle/service/acc/acc.cpp:537]: Perhaps move this to the initializer list?

**Review by Morph1984 — APPROVED** [review]

**Comment by benderscruffy** [comment]
@german77 
did you get it to load in AMD
not loading for me with RX 580

**Comment by german77** [comment]
the game doesn't load but other do. Feel free to open a new issue with a new log since previous one is solved

**Pull request merged**